Add support to pull back all users from Canvas

MakotoTheKnight opened the pull request:
We offer two ways to do this:
 - Through the user search API (which can be slow)
 - Through Account Reports (which can be very, very fast)

We leave the streaming and reading of the account report to any
downstream clients.

There's no hard timeline here to get this merged into Maven Central, as we at CU Boulder are using an in-house fork, but sooner is always appreciated.

kawesome commented:
Can one of the admins verify this patch?

ToeBee commented:
Sorry it has taken me so long to look at this. We had a library fire which resulted in a data center flood that tossed our work schedule into chaos for the summer. And then I went to Florida for a week :)

Anyway, I will need to play with and run this code a little but one comment off the bat I see two `.rej` files that I suspect were accidental commits?

Jenkins: ok to test.

MakotoTheKnight commented:
Right you are @ToeBee; I'll clean that up and resubmit the PR momentarily.

ToeBee commented:
This PR can not be compiled as it stands right now. It looks like you are missing some import statements in a couple of classes and something else is not happy with the arguments to the super constructor. I'm guessing you did a selective commit and missed a few important lines. Please make sure this branch builds from a clean state.

MakotoTheKnight commented:
I'm thinking that the signature for the super constructor may have changed when I wasn't looking.  That's fine; I'll happily resubmit.

ToeBee commented:
Yes, we added a new parameter to all the implementation classes to control the behavior of serializing null fields when sending data to Canvas. See #72 and you will need to apply that change to the constructors in all of your implementation classes or the API factory will be unable to instantiate any concrete classes.

I just tried to give the `AccountReportReader` a test run and discovered that you did not add this class to the `readerMap` in the API factory class

MakotoTheKnight commented:
I do apologize; this certainly wasn't the case on our branch.  I'll be double-checking this once more locally as well.

ToeBee commented:
In addition to still needing to modify the constructor signatures in your implementation classes, I'm thinking that the `AccountReportImpl.listReports` method should check the `report` String being passed in and throw an `IllegalArgumentException` if it is null or an empty String.

Reasoning: I tried calling it with an empty String and the way that Canvas works, that actually ends up executing the `AccountReportSummary.listAvailableReports` API call on the server and passing back different JSON than the `listReports` method is expecting so it ends up returning objects with nothing but nulls in them.

MakotoTheKnight commented:
Apologies for the delay; I've been pulled into another project.  The last commit I had made to this addressed the last piece of feedback you had for me @ToeBee.  I'll triple check later this week against a fresh master line that the signatures for the impl classes are where they should be.

ToeBee commented on src/main/java/edu/ksu/canvas/interfaces/UserWriter.java:
```diff
@@ -25,5 +25,15 @@
+     * @throws IOException When there is an error communicating with Canvas
+     * @see UpdateUserOptions
+     */
+    Optional<User> updateUser(UpdateUserOptions options) throws InvalidOauthTokenException, IOException;
```
Why make this updateUser method when you can set all of these properties on a `User` object and then user the other updateUser method?

ToeBee replied:
Oh, I think I answered my own question. It looks like the existing `updateUser` method doens't actually work correctly. It is only setting two properties instead of serializing the User object to JSON like other update methods do. I should probably look at fixing that...

MakotoTheKnight replied:
Right - I believe at the time I didn't want to cause any potential breakage with existing behavior.  We can definitely unify it into one object once the work to fix up the `updateUser` method is done.

ToeBee commented:
So, I think I fixed the `updateUser` method. The Canvas API was being a little weird with it. When I tried to use the JSON POST method it was throwing a permission error of some kind. I'm guessing it was trying to set some field that isn't allowed to be modified. Most API calls simply ignore extra fields that can't be set. But for some reason the update user one does not. Anyway, I had to use the form post method instead and then realized that this did not support the `serializeNulls` option we added a while ago so I went and did that as well.

Sorry it has taken so long to get this all sorted out. Could you merge from master and see if that `updateUser` method works for you?

MakotoTheKnight commented:
Thanks for looking into this again.  I've moved to a different project but
I'll circle back with my teammates to see if we can get this addressed soon.

On Thu, Nov 29, 2018 at 3:50 PM Toby Murray <notifications@github.com>
wrote:

> So, I think I fixed the updateUser method. The Canvas API was being a
> little weird with it. When I tried to use the JSON POST method it was
> throwing a permission error of some kind. I'm guessing it was trying to set
> some field that isn't allowed to be modified. Most API calls simply ignore
> extra fields that can't be set. But for some reason the update user one
> does not. Anyway, I had to use the form post method instead and then
> realized that this did not support the serializeNulls option we added a
> while ago so I went and did that as well.
>
> Sorry it has taken so long to get this all sorted out. Could you merge
> from master and see if that updateUser method works for you?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kstateome/canvas-api/pull/80#issuecomment-443023421>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvKdoxc3A8CtOWY0JDC1p_beIdE7k4Fks5u0GSjgaJpZM4Ugqm_>
> .
>

MakotoTheKnight commented:
I've talked this over with my team and we'll likely be able to start looking into this around January at the earliest.  We're a bit backed up with a few other projects right now and we can't schedule any new work over this month.

I've rebased my work on top of what you've changed, so I'm at least in an ideal position to start this up again when I can.  Thanks for looking into this again - it's much appreciated!

ToeBee commented:
I went ahead and cleaned things up a little and merged this. I did end up making some pretty substantial changes so your code won't compile against the next release and will need some tweaking.

Thanks for the contributions!
